File: backend/app/analysis/smoothing.py
"""Signal smoothing and gap filling for per-frame time series.

Raw landmark output jitters frame to frame. Rep detection compares a signal
against thresholds, so untreated jitter around a threshold produces phantom
repetitions — smoothing is not cosmetic here, it is what makes the rep counter
trustworthy.

All functions operate on ``list[float | None]``, where None means "not measured
in this frame". They preserve list length so every series stays index-aligned
with the video frames.
"""

from __future__ import annotations

import numpy as np


def window_size_for_fps(fps: float, seconds: float, minimum: int = 3) -> int:
    """Convert a duration into an odd frame count.

    Odd so the window is symmetric about its centre and introduces no phase
    shift — an even window would offset every rep boundary by half a frame in a
    consistent direction, which quietly biases every timing metric.
    """
    if fps <= 0 or seconds <= 0:
        return minimum
    size = int(round(fps * seconds))
    size = max(size, minimum)
    if size % 2 == 0:
        size += 1
    return size


def interpolate_gaps(values: list[float | None], max_gap: int) -> list[float | None]:
    """Linearly fill runs of None no longer than ``max_gap``.

    Short dropouts — a limb briefly occluded — are worth bridging so a rep is
    not split in two. Long dropouts are genuinely missing data and are left as
    None, because inventing a straight line across a second of unseen movement
    would fabricate a rep that may not have happened.

    Leading and trailing gaps are never filled: with data on only one side
    there is nothing to interpolate between, and extrapolating a squat's
    trajectory is guesswork.
    """
    if max_gap <= 0:
        return list(values)

    result = list(values)
    known = [i for i, value in enumerate(result) if value is not None]
    if len(known) < 2:
        return result

    for left, right in zip(known, known[1:], strict=False):
        gap = right - left - 1
        if gap <= 0 or gap > max_gap:
            continue
        start_value = result[left]
        end_value = result[right]
        assert start_value is not None and end_value is not None  # noqa: S101
        step = (end_value - start_value) / (gap + 1)
        for offset in range(1, gap + 1):
            result[left + offset] = start_value + step * offset

    return result


def moving_average(values: list[float | None], window: int) -> list[float | None]:
    """Centred moving average that tolerates missing values.

    Each output is the mean of the present values within the window centred on
    that index. Positions that were None stay None — smoothing must not
    resurrect a frame in which the subject was not tracked, or rep detection
    would run over invented data.

    Windows are truncated at the ends rather than padded. Padding with edge
    values flattens the first and last few frames, which is where a rep often
    starts or finishes.
    """
    if window <= 1 or not values:
        return list(values)

    half = window // 2
    array = np.array(
        [np.nan if value is None else float(value) for value in values],
        dtype=float,
    )
    present = ~np.isnan(array)

    # Cumulative sums give an O(n) sliding window regardless of window size.
    # nan entries contribute zero to the sum and zero to the count, so the mean
    # is automatically taken over present values only.
    filled = np.where(present, array, 0.0)
    sums = np.concatenate(([0.0], np.cumsum(filled)))
    counts = np.concatenate(([0.0], np.cumsum(present.astype(float))))

    n = len(array)
    indices = np.arange(n)
    starts = np.maximum(indices - half, 0)
    ends = np.minimum(indices + half + 1, n)

    window_sums = sums[ends] - sums[starts]
    window_counts = counts[ends] - counts[starts]

    with np.errstate(invalid="ignore", divide="ignore"):
        averaged = np.where(window_counts > 0, window_sums / window_counts, np.nan)

    # Restore the original None mask.
    averaged = np.where(present, averaged, np.nan)
    return [None if np.isnan(value) else float(value) for value in averaged]


def smooth_series(
    values: list[float | None],
    fps: float,
    seconds: float,
    max_gap: int,
) -> list[float | None]:
    """Fill short gaps, then smooth. The standard treatment for every signal.

    Order matters: interpolating first means the moving average sees a
    continuous signal across brief dropouts instead of averaging over a hole.
    """
    return moving_average(
        interpolate_gaps(values, max_gap), window_size_for_fps(fps, seconds)
    )


def percentile(values: list[float | None], q: float) -> float | None:
    """Percentile of the present values, or None if there are none.

    Rep detection uses the 10th and 90th percentiles rather than min and max to
    establish the range of hip travel, because a single mis-tracked frame at
    either extreme would otherwise define the whole scale.
    """
    present = [float(value) for value in values if value is not None]
    if not present:
        return None
    return float(np.percentile(present, q))


def decimate(values: list[float | None], max_points: int) -> list[float | None]:
    """Reduce a series to at most ``max_points`` by uniform sampling.

    Used only when shaping API responses. Sampling rather than averaging keeps
    every returned point a real measurement, so a chart tooltip never shows a
    value that did not occur.
    """
    if max_points <= 0 or len(values) <= max_points:
        return list(values)
    indices = np.linspace(0, len(values) - 1, max_points).round().astype(int)
    return [values[i] for i in indices]


def decimation_indices(length: int, max_points: int) -> list[int]:
    """Indices `decimate` would keep, so parallel series stay aligned."""
    if max_points <= 0 or length <= max_points:
        return list(range(length))
    return [int(i) for i in np.linspace(0, length - 1, max_points).round().astype(int)]

File: frontend/src/lib/analysis/smoothing.ts
/**
 * Signal smoothing and gap filling — the TS mirror of `smoothing.py`.
 *
 * Raw landmarks jitter frame to frame, and rep detection compares a signal
 * against thresholds, so untreated jitter around a threshold invents reps.
 * Every function works on `(number | null)[]` where null means "not measured
 * this frame", and preserves length so series stay index-aligned.
 *
 * This is the batch (centred) treatment, matching the backend exactly, used by
 * `computeAngles` over a full frame array and by the parity tests. The live
 * driver applies a causal variant instead (see `live/`), because a centred
 * window needs frames from the future that a live feed does not have yet.
 */

export type Series = (number | null)[];

/** Convert a duration into an odd frame count (odd = symmetric, no phase shift). */
export function windowSizeForFps(fps: number, seconds: number, minimum = 3): number {
  if (fps <= 0 || seconds <= 0) return minimum;
  let size = Math.max(Math.round(fps * seconds), minimum);
  if (size % 2 === 0) size += 1;
  return size;
}

/**
 * Linearly fill runs of null no longer than `maxGap`. Short dropouts (a briefly
 * occluded limb) are bridged; long ones are left missing rather than fabricated.
 * Leading and trailing gaps are never filled — nothing to interpolate between.
 */
export function interpolateGaps(values: Series, maxGap: number): Series {
  if (maxGap <= 0) return [...values];

  const result = [...values];
  const known: number[] = [];
  result.forEach((value, i) => {
    if (value !== null) known.push(i);
  });
  if (known.length < 2) return result;

  for (let k = 0; k < known.length - 1; k++) {
    const left = known[k];
    const right = known[k + 1];
    const gap = right - left - 1;
    if (gap <= 0 || gap > maxGap) continue;
    const startValue = result[left] as number;
    const endValue = result[right] as number;
    const step = (endValue - startValue) / (gap + 1);
    for (let offset = 1; offset <= gap; offset++) {
      result[left + offset] = startValue + step * offset;
    }
  }
  return result;
}

/**
 * Centred moving average that tolerates missing values. Each output is the mean
 * of the present values in the window centred on that index; positions that were
 * null stay null (smoothing must not resurrect an untracked frame). Windows are
 * truncated at the ends rather than padded.
 */
export function movingAverage(values: Series, window: number): Series {
  if (window <= 1 || values.length === 0) return [...values];

  const half = Math.floor(window / 2);
  const n = values.length;
  const result: Series = new Array(n).fill(null);

  for (let i = 0; i < n; i++) {
    if (values[i] === null) continue;
    let sum = 0;
    let count = 0;
    const start = Math.max(i - half, 0);
    const end = Math.min(i + half, n - 1);
    for (let j = start; j <= end; j++) {
      const value = values[j];
      if (value !== null) {
        sum += value;
        count += 1;
      }
    }
    result[i] = count > 0 ? sum / count : null;
  }
  return result;
}

/** Fill short gaps, then smooth. The standard treatment for every batch signal. */
export function smoothSeries(
  values: Series,
  fps: number,
  seconds: number,
  maxGap: number,
): Series {
  return movingAverage(
    interpolateGaps(values, maxGap),
    windowSizeForFps(fps, seconds),
  );
}

/**
 * Percentile of the present values (linear interpolation, matching numpy's
 * default), or null if there are none. Rep detection uses the 10th/90th rather
 * than min/max so one mis-tracked extreme frame cannot set the whole scale.
 */
export function percentile(values: Series, q: number): number | null {
  const present = values
    .filter((value): value is number => value !== null)
    .sort((a, b) => a - b);
  if (present.length === 0) return null;
  if (present.length === 1) return present[0];

  const rank = (q / 100) * (present.length - 1);
  const lo = Math.floor(rank);
  const hi = Math.ceil(rank);
  if (lo === hi) return present[lo];
  return present[lo] * (hi - rank) + present[hi] * (rank - lo);
}
